filesource/state.go

```diff
@@ -5,16 +5,104 @@ import (
 	"fmt"
 	"time"
 
-	"github.com/sirupsen/logrus"
+	boilerplate "github.com/estuary/connectors/source-boilerplate"
+	pf "github.com/estuary/flow/go/protocols/flow"
+	log "github.com/sirupsen/logrus"
 )
 
-// States is the State of each captured file prefix.
-type States map[string]State
+// bindingStateV1 is the conventional location to persist individual binding states in a top-level
+// JSON object containing the connector state.
+const bindingStateV1 = "bindingStateV1"
 
-func (p States) Validate() error {
-	for prefix, state := range p {
+// unmarshalState for CaptureState provides compatibility with legacy state checkpoints which were a
+// map[string]State keying the stream name to the State. It should be removed when all Filesource
+// captures have migrated to the new state, after which point older captures that haven't migrated
+// will be non-functional.
+func unmarshalState(stateJson json.RawMessage, bindings []*pf.CaptureSpec_Binding) (CaptureState, bool, error) {
```

Comment:
Checking my understanding: this function differs from how the other connectors handle state migration because the old filesource state had stream names as the top-level map keys, and so this function has to first establish whether the state looks like `{"bindingStateV1": ...}` and then if it doesn't the migration logic proceeds much the same as in other connectors.

Reply:
Yep that's right. Added a small additional comment to this migration function which may help future readers as well.